### PR TITLE
Write measurements to influxdb (see issue #201)

### PR DESCRIPTION
**Pull request by mad-ady**
Here is the current implementation. It's functional, but not 100% complete.

### CODE REVIEW

File: lib/Smokeping.pm [hunk 1]
```diff
@@ -3741,6 +3911,14 @@ sub get_config ($$){
     if (not $cfg->{Presentation}{multihost} or not $cfg->{Presentation}{multihost}{colors}){
        $cfg->{Presentation}{multihost}{colors} = "004586 ff420e ffde20 579d1c 7e0021 83caff 314004 aecf00 4b1f6f ff950e c5000b 0084d1";
     }
+    # based on the current configuration - if influxdb support is enabled in the InfluxDB section, load the necessary influxdb perl modules
+    if(defined $cfg->{InfluxDB}{host}){
+	    do_debuglog("Found InfluxDB section. Loading InfluxDB modules");
+	    use InfluxDB::HTTP;
```
Comment (oetiker) [inline lib/Smokeping.pm]: these `use` statements get read at compile time, so you can as well move them to the top of the file.

File: PERL_MODULES [hunk 2]
```diff
@@ -14,3 +14,7 @@ IO::Pty
 LWP
 Authen::Radius
 Path::Tiny
+InfluxDB::HTTP
+InfluxDB::LineProtocol
```
Comment (oetiker) [inline PERL_MODULES]: is this a standalone module, or does it require some influxdb library to be present?

File: lib/Smokeping.pm (hunk 1 above)
Comment (mad-ady) [reply lib/Smokeping.pm]: Thanks for the suggestion. In a previous attempt I wanted to load a module with a precision=s/ms/us value, but it didn't work. I'll move them on top.

File: PERL_MODULES (hunk 2 above)
Comment (mad-ady) [reply PERL_MODULES]: Apart from the modules (and their perl dependencies) there are no other requirements. Communication is done via HTTP POST to the influxdb server. LineProtocol just formats the data in the InfluxDB specification, and the HTTP module is probably a wrapper for LWP, but I haven't looked into it.

File: lib/Smokeping.pm (hunk 1 above)
Comment (mad-ady) [reply lib/Smokeping.pm]: Ok, made the change here: https://github.com/oetiker/SmokePing/pull/215/commits/c4076fcdf9d3ca722df5dbe121bd1a77fac621c7

File: lib/Smokeping.pm [hunk 3]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+            }
+        }
+    }
+
+    #for some reason, InfluxDB::HTTP has a bug and stores 0.000000e+00 as a string, not a float.
+    #this will cause measurement loss in InfluxDB
+    #so, we'll do a dirty hack and convert it to a very small non-zero value
+    # 'U' values are not affected by this (not inserted)
+
+    foreach my $key (keys %idata){
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
    for my $key (sort keys %idata){
```

File: lib/Smokeping.pm [hunk 4]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+        if($idata{$key} == 0){
+            next if($key eq "loss" || $key eq "loss_percent"); #loss was not a float, so no need for this
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
            next if ($key eq "loss" or $key eq "loss_percent"); #loss was not a float, so no need for this
```

File: lib/Smokeping.pm [hunk 5]
```diff
@@ -3916,6 +4088,15 @@ sub load_cfg ($;$) {
                die "ERROR: Could not load Storable Support. This is required for the Charts feature - $@\n" if $@;
            load_sorters $cfg->{Presentation}{charts};
         }
+        #initiate a connection to InfluxDB (if needed)
+        if(! defined $influx && defined $cfg->{'InfluxDB'}{'host'}){
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
        if (not defined $influx and defined $cfg->{InfluxDB}{host}){
```

File: lib/Smokeping.pm [hunk 6]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+    # remove datadir as a prefix
+    $itags{path} = $name;
+    $itags{path}=~s/$cfg->{General}{datadir}//;
+    if ($s ne ""){
+        #master won't have a slave tag value
+        $itags{slave} = $s;
+    }
+
+    #send also probe configuration parameters that are prefixed with influx_. 
+    foreach my $parameter (keys %{$tree}){
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
    for my $parameter (sort keys %$tree){
```

File: lib/Smokeping.pm [hunk 7]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+    if(defined $max && $max ne 'U' ){
+        $idata{"max"} = sprintf("%e", $max);
+    }
+
+
+    $itags{host} = $tree->{host};
+    $itags{title} = $tree->{title};
+    # remove datadir as a prefix
+    $itags{path} = $name;
+    $itags{path}=~s/$cfg->{General}{datadir}//;
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
    $itags{path} =~ s/$cfg->{General}{datadir}//;
```

File: lib/Smokeping.pm [hunk 8]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+    my $max = undef;
+
+    for (1..$pings){
+        if($measurements[${_}-1] ne "U"){
+            $idata{"ping${_}"} = sprintf("%e", $measurements[${_}-1]);
+            $min = $measurements[${_}-1] if($measurements[${_}-1] < $min);
+            $max = $measurements[${_}-1] if($measurements[${_}-1] > $max);
+        }
+    }
+    if($min ne 'U'){
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
    if ($min ne 'U'){
```

File: lib/Smokeping.pm [hunk 9]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+        $idata{"min"} = sprintf("%e", $min);
+    }
+    if(defined $max && $max ne 'U' ){
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
    if (defined $max && $max ne 'U' ){
```

File: lib/Smokeping.pm [hunk 10]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+
+    for (1..$pings){
+        if($measurements[${_}-1] ne "U"){
+            $idata{"ping${_}"} = sprintf("%e", $measurements[${_}-1]);
+            $min = $measurements[${_}-1] if($measurements[${_}-1] < $min);
+            $max = $measurements[${_}-1] if($measurements[${_}-1] > $max);
+        }
+    }
+    if($min ne 'U'){
+        $idata{"min"} = sprintf("%e", $min);
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
        $idata{min} = sprintf('%e', $min);
```

File: lib/Smokeping.pm [hunk 11]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+    $idata{median} = sprintf("%e", $measurements[2]) if($measurements[2] ne "U");
+
+    #skip the first 3 items, since they were processed
+    splice(@measurements, 0, 3);
+
+    my $min = $measurements[1]; #first value
+    my $max = undef;
+
+    for (1..$pings){
+        if($measurements[${_}-1] ne "U"){
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
        if ($measurements[$_] ne "U"){
```

File: lib/Smokeping.pm [hunk 12]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+
+    $idata{median} = sprintf("%e", $measurements[2]) if($measurements[2] ne "U");
+
+    #skip the first 3 items, since they were processed
+    splice(@measurements, 0, 3);
+
+    my $min = $measurements[1]; #first value
+    my $max = undef;
+
+    for (1..$pings){
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
    for (0..$pings-1){
```

File: lib/Smokeping.pm [hunk 13]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+        if($measurements[${_}-1] ne "U"){
+            $idata{"ping${_}"} = sprintf("%e", $measurements[${_}-1]);
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
            $idata{'ping'.(${_}+1)} = sprintf('%e', $measurements[$_]);
```

File: lib/Smokeping.pm [hunk 14]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+            $min = $measurements[${_}-1] if($measurements[${_}-1] < $min);
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
            $min = $measurements[$_] if($measurements[$_] < $min);
```

File: lib/Smokeping.pm [hunk 15]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+            $max = $measurements[${_}-1] if($measurements[${_}-1] > $max);
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
            $max = $measurements[$_] if($measurements[$_] > $max);
```

File: lib/Smokeping.pm [hunk 16]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+    my @measurements = split(/:/, $update->[2]);
+    my $i = 1;
+
+    #Note, we force all measurement data to be float (scientific notation), 
+    #because the data type is derived from the first ever data point, which might be wrong.
+    #in case of measurements with no value (e.g. 'U'), we skip the data point so that influx
+    #knows it's lacking a datapoint and can act accordingly
+
+    #first 3 data points are as follows
+    $idata{uptime} = sprintf("%e", $measurements[0]) if($measurements[0] ne "U");
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
    $idata{uptime} = sprintf('%e', $measurements[0]) if($measurements[0] ne "U");
```

File: lib/Smokeping.pm [hunk 17]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+    #loss is always a number, even when all other are unreachable, so no special treatment
+    my $loss = $measurements[1];
+    $itags{loss} = $loss;
+    $idata{loss} = $loss;
+    #calculate loss as a percentage as well
+    my $loss_percent = int($loss/$pings*100);
+    $itags{loss_percent} = $loss_percent;
+    $idata{loss_percent} = $loss_percent;
+
+    $idata{median} = sprintf("%e", $measurements[2]) if($measurements[2] ne "U");
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
    $idata{median} = sprintf('%e', $measurements[2]) if($measurements[2] ne "U");
```

File: lib/Smokeping.pm [hunk 18]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+    my $name = shift;
+    my $s = shift;
+    my $pings = shift;
+    my $tree = shift;
+    my $update = shift;
+
+    #for a slave cut out the first tilda
+    $s=~s/^~//;
+
+    my @influx_data = ();
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
    my @influx_data;
```

File: lib/Smokeping.pm [hunk 19]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+    my %idata = ();
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
    my %idata;
```

File: lib/Smokeping.pm [hunk 20]
```diff
@@ -2099,6 +2107,122 @@ sub update_rrds($$$$$$) {
...
+    my %itags = ();
```
Comment (oetiker) [inline lib/Smokeping.pm]:
```suggestion
    my %itags;
```

File: lib/Smokeping.pm (hunk 5 above)
Comment (mad-ady) [reply lib/Smokeping.pm]: This was removed entirely with this commit: https://github.com/oetiker/SmokePing/pull/215/commits/c4076fcdf9d3ca722df5dbe121bd1a77fac621c7

**Comment by oetiker** [comment]
thank you!

**Comment by mad-ady** [comment]
Could you comment on this last issue? https://github.com/oetiker/SmokePing/issues/201#issuecomment-584112868
I'd like to do it properly, and not in a hacky way...

**Pull request merged**